Change how Web Inspector extension tabs are loaded.

xeenon opened the pull request:
#### c55c39c2cc604350c1d2a795098d12a1ec8e522b
<pre>
Change how Web Inspector extension tabs are loaded.
<a href="https://bugs.webkit.org/show_bug.cgi?id=243073">https://bugs.webkit.org/show_bug.cgi?id=243073</a>
rdar://86542412

Reviewed by Devin Rousso.
Patch by Kiara Rose &lt;kiara_rose@apple.com&gt;.

Add a dedicated loadURL method to load a URL in a tab instead of relying on evaluateScript.
This avoids so extra indirection and makes is a little easier to debug loading issues.
This change also sets the iframe.src up-front and avoids the empty load bounce.

* Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
(WI.WebInspectorExtensionController.prototype.navigateTabForExtension): Added.
* Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js:
(InspectorFrontendAPI.navigateTabForExtension): Added.
* Source/WebKit/UIProcess/API/APIInspectorExtension.cpp:
(API::InspectorExtension::navigateTab): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
(-[_WKInspector navigateExtensionTabWithIdentifier:toURL:completionHandler:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorExtension.mm:
(-[_WKInspectorExtension navigateToURL:inTabWithIdentifier:completionHandler:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorExtensionHost.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKRemoteWebInspectorViewController.mm:
(-[_WKRemoteWebInspectorViewController navigateExtensionTabWithIdentifier:toURL:completionHandler:]): Added.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIExtensionControllerProxy.cpp:
(WebKit::WebInspectorUIExtensionControllerProxy::navigateTabForExtension): Added.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIExtensionControllerProxy.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.cpp:
(WebKit::WebInspectorUIExtensionController::navigateTabForExtension): Added.
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.messages.in:

Canonical link: <a href="https://commits.webkit.org/252793@main">https://commits.webkit.org/252793@main</a>
</pre>

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
```diff
@@ -197,6 +197,17 @@ WI.WebInspectorExtensionController = class WebInspectorExtensionController exten
         return target.PageAgent.reload.invoke({ignoreCache});
     }
 
+    loadURLForExtension(extensionTabID, sourceURL)
```
NIT: The naming of this makes me think it's gonna do a `fetch` or something.  How about `navigateTabForExtension` (to match `createTabForExtension` naming)?

NIT: `sourceURL` has a kinda specific meaning in Web Inspector (specifically `//# sourceURL=...`).  Can we just use `url`?

xeenon replied:
Using `navigate` makes sense to me.

The ship has sailed for `sourceURL`, it is used in a couple other places in this Inspector Extensions code already.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/WebInspectorExtensionTabContentView.js:
```diff
@@ -78,6 +78,11 @@ WI.WebInspectorExtensionTabContentView = class WebInspectorExtensionTabContentVi
 
+    set iframeURL(sourceURL)
+    {
+        this._iframeElement.src = sourceURL;
```
I'm pretty sure setting the `src` to the same value will still trigger a load.  Do we want that?  Or should we bail if it's the same URL?

xeenon replied:
Yeah, all cases we are calling this we want the extension to load the URL fresh.

dcrousso commented on Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js:
```diff
@@ -245,4 +245,10 @@ InspectorFrontendAPI = {
         return WI.sharedApp.extensionController.evaluateScriptInExtensionTab(extensionTabID, scriptSource);
     },
+
+    // Returns a string (WI.WebInspectorExtension.ErrorCode) if an error occurred that prevented the resource from loading.
```
This comment implies that if the `<iframe>` fails to load (i.e. dispatching an `"error"`) it'd send back the error.  I don't think it's actually doing that.  Should it?  Or should this comment be adjusted?

xeenon replied:
I'll change the comment. The only error returned is if the tab does not exist for that ID.

dcrousso commented on Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
```diff
@@ -290,4 +290,24 @@ - (void)showExtensionTabWithIdentifier:(NSString *)extensionTabIdentifier comple
 #endif
 }
 
+- (void)loadExtensionTabWithIdentifier:(NSString *)extensionTabIdentifier url:(NSURL *)url completionHandler:(void(^)(NSError *))completionHandler
```
ditto (Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:+200) with maybe something like `-navigateTabWithIdentifier:toURL:completionHandler:`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
```diff
@@ -197,6 +197,17 @@ WI.WebInspectorExtensionController = class WebInspectorExtensionController exten
+    {
+        let tabContentView = this._extensionTabContentViewForExtensionTabIDMap.get(extensionTabID);
+        if (!tabContentView) {
+            WI.reportInternalError("Unable to show extension tab with unknown extensionTabID: " + extensionTabID);
```
oops s/show/navigate (or something else)

dcrousso commented on Source/WebKit/UIProcess/API/Cocoa/_WKInspectorExtension.h:
```diff
@@ -75,6 +75,14 @@ WK_CLASS_AVAILABLE(macos(12.0))
+ * @param tabIdentifier Identifier for the Web Inspector tab in which to navigate.
+ * @param completionHandler A block to invoke when the operation completes or fails.
+ */
+- (void)navigateToURL:(NSURL *)url inTabWithIdentifier:(NSString *)tabIdentifier completionHandler:(void(^)(NSError * _Nullable))completionHandler;
```
I feel like `-navigateTabWithIdentifier:toURL:completionHandler:` would be clearer to read and would better match the naming/pattern of `-createTabWithName:tabIconURL:sourceURL:completionHandler:` (and `navigateExtensionTabWithIdentifier:toURL:completionHandler:`)

xeenon replied:
I kept this as-is since it better matches the other methods with `inTabWithIdentifier:` being second.

dcrousso commented on Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
```diff
@@ -290,4 +290,24 @@ - (void)showExtensionTabWithIdentifier:(NSString *)extensionTabIdentifier comple
+        return;
+    }
+
+    _inspector->extensionController()->navigateTabForExtension(extensionTabIdentifier, url, [protectedSelf = retainPtr(self), capturedBlock = makeBlockPtr(completionHandler)] (const std::optional<Inspector::ExtensionError>&& result) mutable {
```
is `protectedSelf` necessary?

xeenon replied:
Yes, for the life of the local `_inspector` used with is implicitly `self->_inspector`.

dcrousso commented on Source/WebKit/UIProcess/API/Cocoa/_WKRemoteWebInspectorViewController.mm:
```diff
@@ -243,6 +243,26 @@ - (void)showExtensionTabWithIdentifier:(NSString *)extensionTabIdentifier comple
+        return;
+    }
+
+    m_remoteInspectorProxy->extensionController()->navigateTabForExtension(extensionTabIdentifier, url, [protectedSelf = retainPtr(self), capturedBlock = makeBlockPtr(completionHandler)] (const std::optional<Inspector::ExtensionError> result) mutable {
```
is `protectedSelf` necessary?

xeenon replied:
Yes, ditto.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/WebInspectorExtensionTabContentView.js:
```diff
@@ -119,13 +124,6 @@ WI.WebInspectorExtensionTabContentView = class WebInspectorExtensionTabContentVi
-        }
-
         // Signal that the page is available since we already bounced to the requested page.
         if (!this._whenPageAvailablePromise.settled)
```
I think we need to reset/recreate this inside `set iframeURL`.  If we leave it as-is, then it probably would negatively impact `WI.WebInspectorExtensionController.prototype.showExtensionTab` because it'd always be resolved after the first load.

xeenon replied:
Always being resolved after the first load is by design, we just care if a page ever loaded for the clients of this promise.

xeenon committed:
Change how Web Inspector extension tabs are loaded.
https://bugs.webkit.org/show_bug.cgi?id=243073
rdar://86542412

Reviewed by Devin Rousso.
Patch by Kiara Rose <kiara_rose@apple.com>.

Add a dedicated loadURL method to load a URL in a tab instead of relying on evaluateScript.
This avoids so extra indirection and makes is a little easier to debug loading issues.
This change also sets the iframe.src up-front and avoids the empty load bounce.

* Source/WebInspectorUI/UserInterface/Controllers/WebInspectorExtensionController.js:
(WI.WebInspectorExtensionController.prototype.navigateTabForExtension): Added.
* Source/WebInspectorUI/UserInterface/Protocol/InspectorFrontendAPI.js:
(InspectorFrontendAPI.navigateTabForExtension): Added.
* Source/WebKit/UIProcess/API/APIInspectorExtension.cpp:
(API::InspectorExtension::navigateTab): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKInspector.mm:
(-[_WKInspector navigateExtensionTabWithIdentifier:toURL:completionHandler:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorExtension.mm:
(-[_WKInspectorExtension navigateToURL:inTabWithIdentifier:completionHandler:]): Added.
* Source/WebKit/UIProcess/API/Cocoa/_WKInspectorExtensionHost.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKRemoteWebInspectorViewController.mm:
(-[_WKRemoteWebInspectorViewController navigateExtensionTabWithIdentifier:toURL:completionHandler:]): Added.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIExtensionControllerProxy.cpp:
(WebKit::WebInspectorUIExtensionControllerProxy::navigateTabForExtension): Added.
* Source/WebKit/UIProcess/Inspector/WebInspectorUIExtensionControllerProxy.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.cpp:
(WebKit::WebInspectorUIExtensionController::navigateTabForExtension): Added.
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.h:
* Source/WebKit/WebProcess/Inspector/WebInspectorUIExtensionController.messages.in:

Canonical link: https://commits.webkit.org/252793@main
